In [31]:
import pandas as pd
import numpy as np

In [2]:
!wget https://raw.githubusercontent.com/BKHV/risk_models/master/data/PD-data-train.csv
!wget https://raw.githubusercontent.com/BKHV/risk_models/master/data/PD-data-test.csv
!wget https://raw.githubusercontent.com/BKHV/risk_models/master/data/PD-data-desc.csv

--2021-11-04 16:03:13--  https://raw.githubusercontent.com/BKHV/risk_models/master/data/PD-data-train.csv
Распознаётся raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.109.133, ...
Подключение к raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... соединение установлено.
HTTP-запрос отправлен. Ожидание ответа... 200 OK
Длина: 5329268 (5,1M) [text/plain]
Сохранение в каталог: ««PD-data-train.csv»».

PD-data-train.csv   100%[===================>]   5,08M  14,1MB/s    за 0,4s    

2021-11-04 16:03:14 (14,1 MB/s) - «PD-data-train.csv» сохранён [5329268/5329268]

--2021-11-04 16:03:14--  https://raw.githubusercontent.com/BKHV/risk_models/master/data/PD-data-test.csv
Распознаётся raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.109.133, ...
Подключение к raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... соединение установлено.
HTTP-запрос от

In [32]:
train_df = pd.read_csv('PD-data-train.csv', sep=';')
test_df = pd.read_csv('PD-data-test.csv', sep=';')
desc_df = pd.read_csv('PD-data-desc.csv', sep=';')

# EDA

### Описание фичей

In [33]:
desc_df

field  \
0                     ar_revenue   
1              ar_total_expenses   
2                   ar_sale_cost   
3            ar_selling_expenses   
4         ar_management_expenses   
5                 ar_sale_profit   
6   ar_balance_of_rvns_and_expns   
7           ar_profit_before_tax   
8                       ar_taxes   
9     ar_other_profit_and_losses   
10                 ar_net_profit   
11         ab_immobilized_assets   
12      ab_mobile_current_assets   
13                  ab_inventory   
14        ab_accounts_receivable   
15       ab_other_current_assets   
16        ab_cash_and_securities   
17                     ab_losses   
18                ab_own_capital   
19           ab_borrowed_capital   
20      ab_long_term_liabilities   
21       ab_short_term_borrowing   
22           ab_accounts_payable   
23           ab_other_borrowings   
24                       bus_age   
25                      ogrn_age   
26                adr_actual_age   
27               head_actual_age   
28                cap_actual_age   
29                ul_staff_range   
30                ul_capital_sum   
31               ul_founders_cnt   
32                 ul_branch_cnt   
33              ul_strategic_flg   
34          ul_systematizing_flg   

                                             desc_eng  \
0         Sales revenue (at the end of the last year)   
1                    Total expenses for the last year   
2         Cost of sales (at the end of the last year)   
3      Selling expenses (at the end of the last year)   
4   Administrative expenses (at the end of the las...   
5     Profit from sales (at the end of the last year)   
6   Balance of income and expenses (at the end of ...   
7                     Gross Profit (end of last year)   
8    Current income tax (at the end of the last year)   
9   Other income and expenses (at the end of the l...   
10           Net profit (at the end of the last year)   
11                     Fixed assets for the last year   
12       Current assets (at the end of the last year)   
13               Stocks (at the end of the last year)   
14  Accounts receivable (at the end of the last year)   
15  Other current assets (at the end of the last y...   
16  Cash and cash equivalents (at the end of the l...   
17    Retained earnings (at the end of the last year)   
18              Capital (at the end of the last year)   
19                      Total debt (end of last year)   
20  Long-term liabilities (at the end of the last ...   
21  Short-term borrowed funds (at the end of the l...   
22                 Accounts payable for the last year   
23    Other liabilities (at the end of the last year)   
24                                  Business duration   
25         Term from the moment the PSRN was assigned   
26  Term from the date of registration of the lega...   
27             Term since the appointment of the head   
28       Term from the moment of capital installation   
29                            The number of employees   
30                                     Capital amount   
31                             Number of shareholders   
32                                 Number of branches   
33                  A company of strategic importance   
34                                   Backbone company   

                                             desc_rus  
0    Выручка от реализации (на конец последнего года)  
1                      Общие расходы за последний год  
2     Себестоимость продаж (на конец последнего года)  
3     Коммерческие расходы (на конец последнего года)  
4   Управленческие расходы (на конец последнего года)  
5    Прибыль от реализации (на конец последнего года)  
6   Баланс доходов и расходов (на конец последнего...  
7          Валовая прибыль (на конец последнего года)  
8   Текущий на налог на прибыль (на конец последне...  
9   Прочие доходы и расходы (на конец последнего г...  
10          Чистая прибыль (на конец последнего года)  
11   

In [34]:
import seaborn as sns
import matplotlib.pyplot as plt

In [35]:
train_df.head()

record_id  ar_revenue  ar_total_expenses  ar_sale_cost  \
0          7  52996000.0         48344000.0    48344000.0   
1         13         NaN                NaN           NaN   
2         18         NaN                NaN           NaN   
3         22  32049000.0         29568000.0    29568000.0   
4         23  16458000.0         14935000.0    14935000.0   

   ar_selling_expenses  ar_management_expenses  ar_sale_profit  \
0                  0.0                     0.0       4652000.0   
1                  NaN                     NaN             NaN   
2                  NaN                     NaN             NaN   
3                  0.0                     0.0       2481000.0   
4                  0.0                     0.0       1523000.0   

   ar_balance_of_rvns_and_expns  ar_profit_before_tax  ar_taxes  ...  \
0                    -2608000.0             2044000.0  846000.0  ...   
1                           NaN                   NaN       NaN  ...   
2                           NaN                   NaN       NaN  ...   
3                     -260000.0             2221000.0  302000.0  ...   
4                     -274000.0             1249000.0  538000.0  ...   

   adr_actual_age  head_actual_age  cap_actual_age  ul_staff_range  \
0               3                3               3         [1-100]   
1              12               12              12         [1-100]   
2              28               28              28         [1-100]   
3               8                8               8         [1-100]   
4              27               27              27         [1-100]   

   ul_capital_sum  ul_founders_cnt  ul_branch_cnt  ul_strategic_flg  \
0         50000.0                2              0                 0   
1         10000.0                5              0                 0   
2         10000.0                2              0                 0   
3        300000.0                1              0                 0   
4         10000.0                2              0                 0   

   ul_systematizing_flg  default_12m  
0                     0            0  
1                     0            0  
2                     0            0  
3                     0            0  
4                     0            0  

[5 rows x 37 columns]

In [36]:
test_df.head()

record_id   ar_revenue  ar_total_expenses  ar_sale_cost  \
0        196   39103000.0         38913000.0    38913000.0   
1       1196          NaN                NaN           NaN   
2       2813   54174000.0         50929000.0    50929000.0   
3       4385    1904000.0          1679000.0     1679000.0   
4       6479  225584000.0        210685000.0   210685000.0   

   ar_selling_expenses  ar_management_expenses  ar_sale_profit  \
0                  0.0                     0.0        190000.0   
1                  NaN                     NaN             NaN   
2                  0.0                     0.0       3245000.0   
3                  0.0                     0.0        225000.0   
4                  0.0                     0.0      14899000.0   

   ar_balance_of_rvns_and_expns  ar_profit_before_tax  ar_taxes  ...  \
0                      141000.0              331000.0  284000.0  ...   
1                           NaN                   NaN       NaN  ...   
2                    -1278000.0             1967000.0  500000.0  ...   
3                      -88000.0              137000.0       0.0  ...   
4                   -12715000.0             2184000.0  445000.0  ...   

   ogrn_age  adr_actual_age  head_actual_age  cap_actual_age  ul_staff_range  \
0       135               3                3               3         [1-100]   
1       138               5                5               5         [1-100]   
2       138              20               20              20         [1-100]   
3       139               9                9               9         [1-100]   
4       139               3                3               3       (100-500]   

   ul_capital_sum  ul_founders_cnt  ul_branch_cnt  ul_strategic_flg  \
0        100000.0                1              0                 0   
1        100000.0                1              0                 0   
2        250000.0                2              0                 0   
3         10000.0                2              0                 0   
4       1700598.0               84              0                 0   

   ul_systematizing_flg  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  

[5 rows x 36 columns]

In [37]:
submit_ids = pd.Series(test_df['record_id'])

Ещё из бейзлайна понятно, что столбец `record_id` нам нужен только для сабмита

In [38]:
train_df = pd.DataFrame(train_df.drop('record_id', axis=1))
test_df = pd.DataFrame(test_df.drop('record_id', axis=1))

In [39]:
train_df.dtypes

ar_revenue                      float64
ar_total_expenses               float64
ar_sale_cost                    float64
ar_selling_expenses             float64
ar_management_expenses          float64
ar_sale_profit                  float64
ar_balance_of_rvns_and_expns    float64
ar_profit_before_tax            float64
ar_taxes                        float64
ar_other_profit_and_losses      float64
ar_net_profit                   float64
ab_immobilized_assets           float64
ab_mobile_current_assets        float64
ab_inventory                    float64
ab_accounts_receivable          float64
ab_other_current_assets         float64
ab_cash_and_securities          float64
ab_losses                       float64
ab_own_capital                  float64
ab_borrowed_capital             float64
ab_long_term_liabilities        float64
ab_short_term_borrowing         float64
ab_accounts_payable             float64
ab_other_borrowings             float64
bus_age                           int64


In [40]:
list(train_df.dtypes)[:-1] == list(test_df.dtypes)  # Не учитывая таргет фичу в трейне

True

In [41]:
desc_df['type'] = list(test_df.dtypes)

In [42]:
desc_df.head()

field                                           desc_eng  \
0              ar_revenue        Sales revenue (at the end of the last year)   
1       ar_total_expenses                   Total expenses for the last year   
2            ar_sale_cost        Cost of sales (at the end of the last year)   
3     ar_selling_expenses     Selling expenses (at the end of the last year)   
4  ar_management_expenses  Administrative expenses (at the end of the las...   

                                            desc_rus     type  
0   Выручка от реализации (на конец последнего года)  float64  
1                     Общие расходы за последний год  float64  
2    Себестоимость продаж (на конец последнего года)  float64  
3    Коммерческие расходы (на конец последнего года)  float64  
4  Управленческие расходы (на конец последнего года)  float64

###### Что с пропусками

In [43]:
train_df.isna().any()

ar_revenue                       True
ar_total_expenses                True
ar_sale_cost                     True
ar_selling_expenses              True
ar_management_expenses           True
ar_sale_profit                   True
ar_balance_of_rvns_and_expns     True
ar_profit_before_tax             True
ar_taxes                         True
ar_other_profit_and_losses       True
ar_net_profit                    True
ab_immobilized_assets            True
ab_mobile_current_assets         True
ab_inventory                     True
ab_accounts_receivable           True
ab_other_current_assets          True
ab_cash_and_securities           True
ab_losses                        True
ab_own_capital                   True
ab_borrowed_capital              True
ab_long_term_liabilities         True
ab_short_term_borrowing          True
ab_accounts_payable              True
ab_other_borrowings              True
bus_age                         False
ogrn_age                        False
adr_actual_a

In [44]:
test_df.isna().any()

ar_revenue                       True
ar_total_expenses                True
ar_sale_cost                     True
ar_selling_expenses              True
ar_management_expenses           True
ar_sale_profit                   True
ar_balance_of_rvns_and_expns     True
ar_profit_before_tax             True
ar_taxes                         True
ar_other_profit_and_losses       True
ar_net_profit                    True
ab_immobilized_assets            True
ab_mobile_current_assets         True
ab_inventory                     True
ab_accounts_receivable           True
ab_other_current_assets          True
ab_cash_and_securities           True
ab_losses                        True
ab_own_capital                   True
ab_borrowed_capital              True
ab_long_term_liabilities         True
ab_short_term_borrowing          True
ab_accounts_payable              True
ab_other_borrowings              True
bus_age                         False
ogrn_age                        False
adr_actual_a

Пропуски есть как в `train_df`, так и в `test_df`

In [45]:
train_df.isna().sum()

ar_revenue                      15145
ar_total_expenses               15145
ar_sale_cost                    15145
ar_selling_expenses             15145
ar_management_expenses          15145
ar_sale_profit                  15145
ar_balance_of_rvns_and_expns    15145
ar_profit_before_tax            15145
ar_taxes                        15145
ar_other_profit_and_losses      15145
ar_net_profit                   15145
ab_immobilized_assets           15145
ab_mobile_current_assets        15145
ab_inventory                    15145
ab_accounts_receivable          15145
ab_other_current_assets         15145
ab_cash_and_securities          15145
ab_losses                       15145
ab_own_capital                  15145
ab_borrowed_capital             15145
ab_long_term_liabilities        15145
ab_short_term_borrowing         15145
ab_accounts_payable             15145
ab_other_borrowings             15145
bus_age                             0
ogrn_age                            0
adr_actual_a

In [46]:
train_df.shape

(32395, 36)

In [47]:
train_df.shape[0] - train_df.isna().sum()[0]

17250

Всего 17250 из 32395 строк в `train_df` не содержат пропуски

In [48]:
(train_df.shape[0] - train_df.isna().sum()[0])/train_df.shape[0]

0.5324895817255749

Это всего 53 процента наблюдений

In [49]:
(train_df.isna().sum() == 0).astype(int).index[(train_df.isna().sum() == 0) == 1]
# фичи без "нанов"

Index(['bus_age', 'ogrn_age', 'adr_actual_age', 'head_actual_age',
       'cap_actual_age', 'ul_staff_range', 'ul_capital_sum', 'ul_founders_cnt',
       'ul_branch_cnt', 'ul_strategic_flg', 'ul_systematizing_flg',
       'default_12m'],
      dtype='object')

In [50]:
(train_df.isna().sum() != 0).value_counts()

True     24
False    12
dtype: int64

Пропуски есть в 24 фичах из 36

In [51]:
train_df.loc[train_df['ar_revenue'].isna()].sample(20)

ar_revenue  ar_total_expenses  ar_sale_cost  ar_selling_expenses  \
30425         NaN                NaN           NaN                  NaN   
13595         NaN                NaN           NaN                  NaN   
13487         NaN                NaN           NaN                  NaN   
28466         NaN                NaN           NaN                  NaN   
6526          NaN                NaN           NaN                  NaN   
11263         NaN                NaN           NaN                  NaN   
16333         NaN                NaN           NaN                  NaN   
15692         NaN                NaN           NaN                  NaN   
16142         NaN                NaN           NaN                  NaN   
477           NaN                NaN           NaN                  NaN   
23339         NaN                NaN           NaN                  NaN   
8636          NaN                NaN           NaN                  NaN   
2040          NaN                NaN           NaN                  NaN   
20359         NaN                NaN           NaN                  NaN   
5378          NaN                NaN           NaN                  NaN   
7764          NaN                NaN           NaN                  NaN   
1801          NaN                NaN           NaN                  NaN   
9119          NaN                NaN           NaN                  NaN   
30520         NaN                NaN           NaN                  NaN   
23865         NaN                NaN           NaN                  NaN   

       ar_management_expenses  ar_sale_profit  ar_balance_of_rvns_and_expns  \
30425                     NaN             NaN                           NaN   
13595                     NaN             NaN                           NaN   
13487                     NaN             NaN                           NaN   
28466                     NaN             NaN                           NaN   
6526                      NaN             NaN                           NaN   
11263                     NaN             NaN                           NaN   
16333                     NaN             NaN                           NaN   
15692                     NaN             NaN                           NaN   
16142                     NaN             NaN                           NaN   
477                       NaN             NaN                           NaN   
23339                     NaN             NaN                           NaN   
8636                      NaN             NaN                           NaN   
2040                      NaN             NaN                           NaN   
20359                     NaN             NaN                           NaN   
5378                      NaN             NaN                           NaN   
7764                      NaN             NaN                           NaN   
1801                      NaN             NaN                           NaN   
9119                      NaN             NaN                           NaN   
30520                     NaN             NaN                           NaN   
23865                     NaN             NaN                           NaN   

       ar_profit_before_tax  ar_taxes  ar_other_profit_and_losses  ...  \
30425                   NaN       NaN                         NaN  ...   
13595                   NaN       NaN                         NaN  ...   
13487                   NaN       NaN                         NaN  ...   
28466                   NaN       NaN                         NaN  ...   
6526                    NaN       NaN                         NaN  ...   
11263                   NaN       NaN                         NaN  ...   
16333                   NaN       NaN                         NaN  ...   
15692                   NaN       NaN                         NaN  ...   
16142                   NaN       NaN                         NaN  ...   
477                     NaN    

Если у наблюдения есть пропуск, то у него пропущены все 24 первые фичи, остальные 12 - на месте 

Посмотрим на `test_df`

In [52]:
test_df.isna().sum()

ar_revenue                      118
ar_total_expenses               118
ar_sale_cost                    118
ar_selling_expenses             118
ar_management_expenses          118
ar_sale_profit                  118
ar_balance_of_rvns_and_expns    118
ar_profit_before_tax            118
ar_taxes                        118
ar_other_profit_and_losses      118
ar_net_profit                   118
ab_immobilized_assets           118
ab_mobile_current_assets        118
ab_inventory                    118
ab_accounts_receivable          118
ab_other_current_assets         118
ab_cash_and_securities          118
ab_losses                       118
ab_own_capital                  118
ab_borrowed_capital             118
ab_long_term_liabilities        118
ab_short_term_borrowing         118
ab_accounts_payable             118
ab_other_borrowings             118
bus_age                           0
ogrn_age                          0
adr_actual_age                    0
head_actual_age             

In [53]:
(test_df.isna().sum()[0])/test_df.shape[0]

0.59

Пропуски есть в 118 наблюдениях из 200 (59 процентов наблюдений)

In [54]:
test_df.loc[test_df['ar_revenue'].isna()].sample(20)

ar_revenue  ar_total_expenses  ar_sale_cost  ar_selling_expenses  \
24          NaN                NaN           NaN                  NaN   
119         NaN                NaN           NaN                  NaN   
50          NaN                NaN           NaN                  NaN   
64          NaN                NaN           NaN                  NaN   
82          NaN                NaN           NaN                  NaN   
110         NaN                NaN           NaN                  NaN   
40          NaN                NaN           NaN                  NaN   
178         NaN                NaN           NaN                  NaN   
97          NaN                NaN           NaN                  NaN   
14          NaN                NaN           NaN                  NaN   
133         NaN                NaN           NaN                  NaN   
27          NaN                NaN           NaN                  NaN   
22          NaN                NaN           NaN                  NaN   
153         NaN                NaN           NaN                  NaN   
145         NaN                NaN           NaN                  NaN   
75          NaN                NaN           NaN                  NaN   
125         NaN                NaN           NaN                  NaN   
196         NaN                NaN           NaN                  NaN   
139         NaN                NaN           NaN                  NaN   
70          NaN                NaN           NaN                  NaN   

     ar_management_expenses  ar_sale_profit  ar_balance_of_rvns_and_expns  \
24                      NaN             NaN                           NaN   
119                     NaN             NaN                           NaN   
50                      NaN             NaN                           NaN   
64                      NaN             NaN                           NaN   
82                      NaN             NaN                           NaN   
110                     NaN             NaN                           NaN   
40                      NaN             NaN                           NaN   
178                     NaN             NaN                           NaN   
97                      NaN             NaN                           NaN   
14                      NaN             NaN                           NaN   
133                     NaN             NaN                           NaN   
27                      NaN             NaN                           NaN   
22                      NaN             NaN                           NaN   
153                     NaN             NaN                           NaN   
145                     NaN             NaN                           NaN   
75                      NaN             NaN                           NaN   
125                     NaN             NaN                           NaN   
196                     NaN             NaN                           NaN   
139                     NaN             NaN                           NaN   
70                      NaN             NaN                           NaN   

     ar_profit_before_tax  ar_taxes  ar_other_profit_and_losses  ...  \
24                    NaN       NaN                         NaN  ...   
119                   NaN       NaN                         NaN  ...   
50                    NaN       NaN                         NaN  ...   
64                    NaN       NaN                         NaN  ...   
82                    NaN       NaN                         NaN  ...   
110                   NaN       NaN                         NaN  ...   
40                    NaN       NaN                         NaN  ...   
178                   NaN       NaN                         NaN  ...   
97                    NaN       NaN                         NaN  ...   
14                    NaN       NaN                         NaN  ...   
133                   NaN       NaN                         N

Ситуация та же, что и с трейном. Если есть пропуск, значит это пропуск в 24 фичах сразу

###### Смотрим на целевую переменную

In [55]:
train_df['default_12m'].value_counts()

0    30304
1     2091
Name: default_12m, dtype: int64

In [56]:
train_df['default_12m'].value_counts(normalize=True)

0    0.935453
1    0.064547
Name: default_12m, dtype: float64

Всего лишь у 6% наблюдений целевая переменная равна 1. Явный дисбаланс

А какой процент будет у наблюдений без пропусков

In [57]:
train_df.loc[train_df['ar_revenue'].notna()]['default_12m'].value_counts(normalize=True)

0    0.953855
1    0.046145
Name: default_12m, dtype: float64

Если мы урежем наш датасет до 17 тысяч наблюдений (наблюдения только без пропусков), то среди них только у 4 с половиной процента целевая фича равна 1, что ещё хуже, чем при полном датасете

# Model

Первое, что мне пришло в голову - построить ансамбль простых классификаторов на выборке с наблюдениями и фичами без пропусков

In [61]:
crop_train_df = pd.DataFrame(train_df.loc[:, 
                                          (train_df.isna().sum() == 0).astype(int)
                                          .index[(train_df.isna().sum() == 0) == 1]])

In [62]:
crop_test_df = pd.DataFrame(test_df.loc[:, 
                                          (test_df.isna().sum() == 0).astype(int)
                                          .index[(test_df.isna().sum() == 0) == 1]])

`crop_train_df` и `crop_test_df` - таблицы без фичей, в которых есть наны. кол-во строк то же

In [63]:
dummies = pd.get_dummies(train_df['ul_staff_range'])

In [64]:
dummies_2 = pd.get_dummies(test_df['ul_staff_range'])

In [65]:
train_df = pd.DataFrame(train_df.drop('ul_staff_range', axis=1))
test_df = pd.DataFrame(test_df.drop('ul_staff_range', axis=1))
crop_train_df = pd.DataFrame(crop_train_df.drop('ul_staff_range', axis=1))
crop_test_df = pd.DataFrame(crop_test_df.drop('ul_staff_range', axis=1))

train_df = pd.concat([train_df, dummies], axis=1)
test_df = pd.concat([test_df, dummies_2], axis=1)
crop_train_df = pd.concat([crop_train_df, dummies], axis=1)
crop_test_df = pd.concat([crop_test_df, dummies_2], axis=1)

In [66]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV

In [67]:
X_train, X_test, y_train, y_test = train_test_split(crop_train_df.drop('default_12m', axis=1), 
                                                    crop_train_df['default_12m'],
                                                    random_state=17, shuffle=True)

In [69]:
clf = RandomForestClassifier(n_jobs=-1)
rf_cv_score = cross_val_score(clf, X_train, y_train, cv=5, scoring='roc_auc').mean()
rf_cv_score

0.6012954894210649

In [70]:
from sklearn.linear_model import LogisticRegression
logit = LogisticRegression(C=1,solver='lbfgs',max_iter=1000, n_jobs=-1)
log_cv_score = cross_val_score(logit, X_train, y_train, cv=5, scoring='roc_auc').mean()
log_cv_score

0.6505874958657077

In [72]:
%%time

forest_params = {'n_estimators': [100, 150, 250, 400],
                 'max_depth': [5, 10, 20, 35, 50],
                 'min_samples_split': [2, 5, 10]}

best_tree = GridSearchCV(RandomForestClassifier(random_state=42, n_jobs=-1),
                                               forest_params, cv=5, n_jobs=-1)

best_tree.fit(X_train, y_train)

CPU times: user 7.86 s, sys: 631 ms, total: 8.49 s
Wall time: 3min 37s


GridSearchCV(cv=5, estimator=RandomForestClassifier(n_jobs=-1, random_state=42),
             n_jobs=-1,
             param_grid={'max_depth': [5, 10, 20, 35, 50],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [100, 150, 250, 400]})

In [73]:
best_tree.best_estimator_

RandomForestClassifier(max_depth=10, n_estimators=250, n_jobs=-1,
                       random_state=42)

In [74]:
best_tree.best_params_

{'max_depth': 10, 'min_samples_split': 2, 'n_estimators': 250}

In [75]:
from sklearn.model_selection import cross_val_predict
best_rf_cv_score = cross_val_score(RandomForestClassifier(random_state=42, n_jobs=-1, max_depth=10, min_samples_split=2, n_estimators=250), X_train, y_train, cv=5, scoring='roc_auc').mean()
best_rf_cv_score

0.6715115106147833

In [76]:
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.pipeline import make_pipeline

In [86]:
%%time

logit_params = {'C': [0.8, 1, 1.2],
               'solver': ['lbfgs', 'sag', 'saga'],
               'max_iter': [100, 125, 150]}
best_logit = GridSearchCV(LogisticRegression(n_jobs=-1), logit_params,
                         cv=5, n_jobs=-1)
best_logit.fit(X_train, y_train)

CPU times: user 723 ms, sys: 215 ms, total: 938 ms
Wall time: 25.7 s


GridSearchCV(cv=5, estimator=LogisticRegression(n_jobs=-1), n_jobs=-1,
             param_grid={'C': [0.8, 1, 1.2], 'max_iter': [100, 125, 150],
                         'solver': ['lbfgs', 'sag', 'saga']})

In [87]:
best_logit.best_params_

{'C': 0.8, 'max_iter': 100, 'solver': 'lbfgs'}

In [88]:
best_logit_cv_score = cross_val_score(LogisticRegression(C=0.8, solver='lbfgs', max_iter=100), X_train, y_train, cv=5, scoring='roc_auc').mean()
best_logit_cv_score

0.6507239622411405

In [89]:
%%time

from sklearn.preprocessing import StandardScaler

pipe = make_pipeline(StandardScaler(), SVC(gamma='auto'))
svm_cv_score = cross_val_score(pipe, X_train, y_train, cv=5, scoring='roc_auc').mean()
svm_cv_score

CPU times: user 16.1 s, sys: 988 ms, total: 17.1 s
Wall time: 16.4 s


0.5589631682797815

In [90]:
%%time

bayes_params = {'var_smoothing': [1e-10, 1e-9, 1e-8, 1e-7, 1e-6]}
best_bayes = GridSearchCV(GaussianNB, bayes_params, n_jobs=-1, cv=5, scoring='accuracy')

CPU times: user 0 ns, sys: 68 µs, total: 68 µs
Wall time: 76.8 µs


In [91]:
rf = RandomForestClassifier(max_depth=10, min_samples_split=10, n_estimators=250, n_jobs=-1)
lr = LogisticRegression(solver='lbfgs', C=0.8, max_iter=100)
ga = GaussianNB()
svm = make_pipeline(StandardScaler(), SVC(gamma='auto', probability=True))

main_clf = VotingClassifier(estimators=[
    ('rf', rf),
    ('lr', lr),
    ('gauss', ga),
    ('SVM', svm)
], voting='soft', n_jobs=-1, verbose=True)

In [92]:
main_clf.fit(crop_train_df.drop('default_12m', axis=1), crop_train_df['default_12m'])

VotingClassifier(estimators=[('rf',
                              RandomForestClassifier(max_depth=10,
                                                     min_samples_split=10,
                                                     n_estimators=250,
                                                     n_jobs=-1)),
                             ('lr', LogisticRegression(C=0.8)),
                             ('gauss', GaussianNB()),
                             ('SVM',
                              Pipeline(steps=[('standardscaler',
                                               StandardScaler()),
                                              ('svc',
                                               SVC(gamma='auto',
                                                   probability=True))]))],
                 n_jobs=-1, verbose=True, voting='soft')

In [93]:
main_clf.predict(crop_test_df)

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0])

И после предсказания я получил все нули :(  


И я залил их на сайт. Мой accuracy был 0.505 или (101/200), это значит, что в тестовой выборке у 101 наблюдения действительно целевая переменная равна 0. У остальных 99 наблюдений она равна 1.  
Примечательно, что соотношение в тестовой выборке почти 1 к 1. В обучающей же почти 1 к 20

Теперь я хочу самые неуверенные из этих нулей превратить в единицы

In [108]:
main_clf_probs = main_clf.predict_proba(crop_test_df.iloc[:])

In [147]:
(main_clf_probs[:, 0] > 0.82).sum()

121

Классификатор только для 121 наблюдений уверен более чем на 82 процента в том, что это наблюдение ноль. Попробуем остальным 79 наблюдениям поставить единицы

In [148]:
main_clf_probs[:, 0] <= 0.82

array([False, False, False, False, False, False, False, False,  True,
        True,  True, False, False, False,  True, False, False, False,
        True, False,  True,  True,  True, False, False,  True, False,
        True, False, False, False, False, False, False, False, False,
       False,  True, False, False, False,  True, False, False, False,
       False, False, False, False,  True,  True,  True,  True, False,
        True,  True, False,  True,  True, False, False, False,  True,
       False, False,  True, False, False, False,  True, False, False,
        True,  True,  True, False, False,  True,  True,  True,  True,
        True,  True, False,  True,  True, False, False, False, False,
        True, False,  True,  True, False, False, False,  True,  True,
       False, False,  True,  True,  True, False,  True, False,  True,
       False,  True,  True, False, False, False, False,  True, False,
        True,  True, False,  True,  True,  True, False, False,  True,
       False, False,

In [151]:
filtered_main_clf_preds = (main_clf_probs[:, 0] <= 0.82)

In [152]:
filtered_main_clf_preds = filtered_main_clf_preds.astype(int)

In [153]:
sudmit_df = pd.DataFrame.from_dict({'id': submit_ids, 'predict': filtered_main_clf_preds})

In [154]:
sudmit_df['predict'].value_counts()

0    121
1     79
Name: predict, dtype: int64

In [155]:
sudmit_df.to_csv('PD-submit.csv', index=False, sep=';')

`accuracy` = ___0.595___ (119 из 200 правильных ответов)

Что если мы снизим порог до 0.775

In [161]:
filtered_main_clf_preds = (main_clf_probs[:, 0] <= 0.775)
filtered_main_clf_preds = filtered_main_clf_preds.astype(int)
sudmit_df = pd.DataFrame.from_dict({'id': submit_ids, 'predict': filtered_main_clf_preds})
sudmit_df['predict'].value_counts()

0    167
1     33
Name: predict, dtype: int64

In [162]:
sudmit_df.to_csv('PD-submit.csv', index=False, sep=';')

Лучший `accuracy` остался тем же (0.595)

Попробуем еще два раза поиграть с этими порогами

In [167]:
filtered_main_clf_preds = (main_clf_probs[:, 0] <= 0.8)
filtered_main_clf_preds = filtered_main_clf_preds.astype(int)
sudmit_df = pd.DataFrame.from_dict({'id': submit_ids, 'predict': filtered_main_clf_preds})
sudmit_df['predict'].value_counts()

0    143
1     57
Name: predict, dtype: int64

In [168]:
sudmit_df.to_csv('PD-submit.csv', index=False, sep=';')

Лучший `accuracy` остался тем же (0.595)

In [172]:
filtered_main_clf_preds = (main_clf_probs[:, 0] <= 0.8425)
filtered_main_clf_preds = filtered_main_clf_preds.astype(int)
submit_df = pd.DataFrame.from_dict({'id': submit_ids, 'predict': filtered_main_clf_preds})
submit_df['predict'].value_counts()

0    100
1    100
Name: predict, dtype: int64

In [173]:
submit_df.to_csv('PD-submit.csv', index=False, sep=';')

In [174]:
print("you've stopped here")

you've stopped here


In [58]:
from catboost import CatBoostClassifier, Pool

In [59]:
model = CatBoostClassifier(learning_rate=0.2, cat_features=[29])
model.fit(train_df.drop('default_12m', axis=1), train_df['default_12m'], plot=True)

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

0:	learn: 0.4847465	total: 24.7ms	remaining: 24.7s
1:	learn: 0.3716480	total: 49.5ms	remaining: 24.7s
2:	learn: 0.3101748	total: 85.2ms	remaining: 28.3s
3:	learn: 0.2754140	total: 116ms	remaining: 28.8s
4:	learn: 0.2563337	total: 139ms	remaining: 27.7s
5:	learn: 0.2450426	total: 162ms	remaining: 26.8s
6:	learn: 0.2372035	total: 193ms	remaining: 27.4s
7:	learn: 0.2320102	total: 219ms	remaining: 27.2s
8:	learn: 0.2290380	total: 246ms	remaining: 27.1s
9:	learn: 0.2273046	total: 277ms	remaining: 27.4s
10:	learn: 0.2252418	total: 302ms	remaining: 27.2s
11:	learn: 0.2242544	total: 327ms	remaining: 27s
12:	learn: 0.2233695	total: 354ms	remaining: 26.9s
13:	learn: 0.2223833	total: 375ms	remaining: 26.4s
14:	learn: 0.2217225	total: 395ms	remaining: 26s
15:	learn: 0.2207877	total: 414ms	remaining: 25.5s
16:	learn: 0.2200364	total: 434ms	remaining: 25.1s
17:	learn: 0.2193034	total: 457ms	remaining: 24.9s
18:	learn: 0.2187671	total: 479ms	remaining: 24.7s
19:	learn: 0.2184815	total: 503ms	remainin

168:	learn: 0.1836490	total: 4.05s	remaining: 19.9s
169:	learn: 0.1834628	total: 4.08s	remaining: 19.9s
170:	learn: 0.1831824	total: 4.1s	remaining: 19.9s
171:	learn: 0.1829870	total: 4.13s	remaining: 19.9s
172:	learn: 0.1828796	total: 4.15s	remaining: 19.8s
173:	learn: 0.1826660	total: 4.17s	remaining: 19.8s
174:	learn: 0.1825873	total: 4.19s	remaining: 19.8s
175:	learn: 0.1823998	total: 4.21s	remaining: 19.7s
176:	learn: 0.1821134	total: 4.23s	remaining: 19.7s
177:	learn: 0.1818581	total: 4.26s	remaining: 19.7s
178:	learn: 0.1818282	total: 4.28s	remaining: 19.6s
179:	learn: 0.1814053	total: 4.31s	remaining: 19.6s
180:	learn: 0.1812694	total: 4.33s	remaining: 19.6s
181:	learn: 0.1810010	total: 4.36s	remaining: 19.6s
182:	learn: 0.1808495	total: 4.38s	remaining: 19.6s
183:	learn: 0.1806397	total: 4.4s	remaining: 19.5s
184:	learn: 0.1804109	total: 4.43s	remaining: 19.5s
185:	learn: 0.1803612	total: 4.47s	remaining: 19.5s
186:	learn: 0.1800558	total: 4.5s	remaining: 19.6s
187:	learn: 0.1

333:	learn: 0.1615123	total: 7.67s	remaining: 15.3s
334:	learn: 0.1614230	total: 7.7s	remaining: 15.3s
335:	learn: 0.1613486	total: 7.72s	remaining: 15.3s
336:	learn: 0.1612318	total: 7.75s	remaining: 15.2s
337:	learn: 0.1611490	total: 7.77s	remaining: 15.2s
338:	learn: 0.1610296	total: 7.79s	remaining: 15.2s
339:	learn: 0.1608790	total: 7.81s	remaining: 15.2s
340:	learn: 0.1607783	total: 7.83s	remaining: 15.1s
341:	learn: 0.1605574	total: 7.85s	remaining: 15.1s
342:	learn: 0.1605028	total: 7.87s	remaining: 15.1s
343:	learn: 0.1604872	total: 7.89s	remaining: 15.1s
344:	learn: 0.1603853	total: 7.93s	remaining: 15.1s
345:	learn: 0.1603109	total: 7.95s	remaining: 15s
346:	learn: 0.1601641	total: 7.97s	remaining: 15s
347:	learn: 0.1601222	total: 7.99s	remaining: 15s
348:	learn: 0.1600861	total: 8.01s	remaining: 14.9s
349:	learn: 0.1599620	total: 8.03s	remaining: 14.9s
350:	learn: 0.1599264	total: 8.05s	remaining: 14.9s
351:	learn: 0.1597756	total: 8.07s	remaining: 14.9s
352:	learn: 0.15952

497:	learn: 0.1468163	total: 11.3s	remaining: 11.4s
498:	learn: 0.1467560	total: 11.3s	remaining: 11.4s
499:	learn: 0.1466747	total: 11.3s	remaining: 11.3s
500:	learn: 0.1465448	total: 11.4s	remaining: 11.3s
501:	learn: 0.1464840	total: 11.4s	remaining: 11.3s
502:	learn: 0.1464601	total: 11.4s	remaining: 11.3s
503:	learn: 0.1463925	total: 11.4s	remaining: 11.3s
504:	learn: 0.1462538	total: 11.5s	remaining: 11.2s
505:	learn: 0.1462453	total: 11.5s	remaining: 11.2s
506:	learn: 0.1462352	total: 11.5s	remaining: 11.2s
507:	learn: 0.1462091	total: 11.5s	remaining: 11.2s
508:	learn: 0.1461181	total: 11.6s	remaining: 11.1s
509:	learn: 0.1460335	total: 11.6s	remaining: 11.1s
510:	learn: 0.1458882	total: 11.6s	remaining: 11.1s
511:	learn: 0.1458143	total: 11.6s	remaining: 11.1s
512:	learn: 0.1456747	total: 11.6s	remaining: 11s
513:	learn: 0.1456315	total: 11.7s	remaining: 11s
514:	learn: 0.1454619	total: 11.7s	remaining: 11s
515:	learn: 0.1453858	total: 11.7s	remaining: 11s
516:	learn: 0.145288

660:	learn: 0.1353059	total: 14.9s	remaining: 7.64s
661:	learn: 0.1352289	total: 14.9s	remaining: 7.62s
662:	learn: 0.1351997	total: 14.9s	remaining: 7.6s
663:	learn: 0.1351633	total: 15s	remaining: 7.58s
664:	learn: 0.1351464	total: 15s	remaining: 7.55s
665:	learn: 0.1350770	total: 15s	remaining: 7.53s
666:	learn: 0.1349983	total: 15s	remaining: 7.51s
667:	learn: 0.1349382	total: 15.1s	remaining: 7.48s
668:	learn: 0.1348838	total: 15.1s	remaining: 7.46s
669:	learn: 0.1348635	total: 15.1s	remaining: 7.44s
670:	learn: 0.1348196	total: 15.1s	remaining: 7.42s
671:	learn: 0.1347666	total: 15.2s	remaining: 7.39s
672:	learn: 0.1346304	total: 15.2s	remaining: 7.37s
673:	learn: 0.1346142	total: 15.2s	remaining: 7.35s
674:	learn: 0.1345692	total: 15.2s	remaining: 7.33s
675:	learn: 0.1345469	total: 15.2s	remaining: 7.3s
676:	learn: 0.1344919	total: 15.3s	remaining: 7.28s
677:	learn: 0.1344074	total: 15.3s	remaining: 7.25s
678:	learn: 0.1343793	total: 15.3s	remaining: 7.23s
679:	learn: 0.1342608	

821:	learn: 0.1271817	total: 18.5s	remaining: 4s
822:	learn: 0.1271371	total: 18.5s	remaining: 3.98s
823:	learn: 0.1271183	total: 18.5s	remaining: 3.96s
824:	learn: 0.1270570	total: 18.6s	remaining: 3.94s
825:	learn: 0.1270367	total: 18.6s	remaining: 3.91s
826:	learn: 0.1269889	total: 18.6s	remaining: 3.89s
827:	learn: 0.1269618	total: 18.6s	remaining: 3.87s
828:	learn: 0.1269331	total: 18.6s	remaining: 3.84s
829:	learn: 0.1268874	total: 18.7s	remaining: 3.82s
830:	learn: 0.1268018	total: 18.7s	remaining: 3.8s
831:	learn: 0.1267646	total: 18.7s	remaining: 3.77s
832:	learn: 0.1267290	total: 18.7s	remaining: 3.75s
833:	learn: 0.1266743	total: 18.7s	remaining: 3.73s
834:	learn: 0.1266589	total: 18.8s	remaining: 3.71s
835:	learn: 0.1266473	total: 18.8s	remaining: 3.69s
836:	learn: 0.1265439	total: 18.8s	remaining: 3.66s
837:	learn: 0.1265106	total: 18.8s	remaining: 3.64s
838:	learn: 0.1264896	total: 18.8s	remaining: 3.62s
839:	learn: 0.1264813	total: 18.9s	remaining: 3.59s
840:	learn: 0.12

989:	learn: 0.1190732	total: 21.9s	remaining: 221ms
990:	learn: 0.1190627	total: 21.9s	remaining: 199ms
991:	learn: 0.1190306	total: 21.9s	remaining: 177ms
992:	learn: 0.1190250	total: 22s	remaining: 155ms
993:	learn: 0.1190075	total: 22s	remaining: 133ms
994:	learn: 0.1189513	total: 22s	remaining: 111ms
995:	learn: 0.1189047	total: 22s	remaining: 88.5ms
996:	learn: 0.1188644	total: 22s	remaining: 66.3ms
997:	learn: 0.1188376	total: 22.1s	remaining: 44.2ms
998:	learn: 0.1188373	total: 22.1s	remaining: 22.1ms
999:	learn: 0.1187558	total: 22.1s	remaining: 0us


In [60]:
init_cat_predictions = model.predict(test_df)

In [232]:
import random

In [234]:
crop_train_df[crop_train_df['default_12m']==0].sample(10000)

bus_age  ogrn_age  adr_actual_age  head_actual_age  cap_actual_age  \
16594      262       136              17               17              17   
25155       45        45               3                3               3   
22943      129       129              18               18              18   
28259      234       139               7                7               7   
4770        69        69              11               11              11   
...        ...       ...             ...              ...             ...   
28536       94        94              30               30              30   
23163       61        61               3                3               3   
24844       46        46               7                7               7   
28799      135       135              23               23              23   
6851        74        74              27               27              27   

       ul_capital_sum  ul_founders_cnt  ul_branch_cnt  ul_strategic_flg  \
16594     303999400.0              768              0                 0   
25155         10000.0                1              0                 0   
22943         11000.0                2              0                 0   
28259       6843200.0                8              0                 0   
4770          10000.0                1              0                 0   
...               ...              ...            ...               ...   
28536         10000.0                1              0                 0   
23163         10000.0                4              0                 0   
24844         10000.0                2              0                 0   
28799         10000.0                1              0                 0   
6851          10000.0                1              0                 0   

       ul_systematizing_flg  default_12m  (100-500]  > 500  [1-100]  
16594                     0            0          1      0        0  
25155                     0            0          0      0        1  
22943                     0            0          0      0        1  
28259                     0            0          0      0        1  
4770                      0            0          0      0        1  
...                     ...          ...        ...    ...      ...  
28536                     0            0          0      0        1  
23163                     0            0          0      0        1  
24844                     0            0          0      0        1  
28799                     0            0          0      0        1  
6851                      0            0          0      0        1  

[10000 rows x 14 columns]

[Voting] .................... (3 of 4) Processing gauss, total=   0.0s
[Voting] ....................... (1 of 4) Processing rf, total=   1.2s
[Voting] .................... (3 of 4) Processing gauss, total=   0.0s
[Voting] ....................... (2 of 4) Processing lr, total=   0.1s
[Voting] ....................... (2 of 4) Processing lr, total=   0.1s
[Voting] ....................... (2 of 4) Processing lr, total=   0.3s
[Voting] ....................... (1 of 4) Processing rf, total=   1.2s
[Voting] .................... (3 of 4) Processing gauss, total=   0.0s
[Voting] ....................... (1 of 4) Processing rf, total=   1.6s


In [235]:
test_0 = crop_train_df[crop_train_df['default_12m']==0].sample(10000)

[Voting] ...................... (4 of 4) Processing SVM, total=  23.2s
[Voting] ...................... (4 of 4) Processing SVM, total=  23.0s
[Voting] ...................... (4 of 4) Processing SVM, total=  47.4s


In [236]:
test_1 = crop_train_df[crop_train_df['default_12m'] == 1]

In [237]:
test = pd.concat([test_0, test_1])

In [239]:
test['default_12m'].value_counts(normalize=True)

0    0.827061
1    0.172939
Name: default_12m, dtype: float64

In [246]:
%%time

main_clf.fit(test.drop('default_12m', axis=1), test['default_12m'])

CPU times: user 63.1 ms, sys: 0 ns, total: 63.1 ms
Wall time: 14 s


VotingClassifier(estimators=[('rf',
                              RandomForestClassifier(max_depth=10,
                                                     min_samples_split=10,
                                                     n_estimators=250,
                                                     n_jobs=-1)),
                             ('lr', LogisticRegression(C=0.8)),
                             ('gauss', GaussianNB()),
                             ('SVM',
                              Pipeline(steps=[('standardscaler',
                                               StandardScaler()),
                                              ('svc',
                                               SVC(gamma='auto',
                                                   probability=True))]))],
                 n_jobs=-1, verbose=True, voting='soft', weights=[4, 1, 1, 1])

In [247]:
preds = main_clf.predict(crop_test_df)

In [248]:
preds.shape

(200,)

In [249]:
preds

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0])

In [244]:
rf.fit(test.drop('default_12m', axis=1), test['default_12m'])
pr = rf.predict(crop_test_df)

pr

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0])

In [250]:
kek = pd.read_csv('PD-data-test.csv', sep=';')

In [251]:
kek['record_id']

0         196
1        1196
2        2813
3        4385
4        6479
        ...  
195    265577
196    266648
197    266849
198    267566
199    271584
Name: record_id, Length: 200, dtype: int64

[Voting] ....................... (2 of 4) Processing lr, total=   0.0s
[Voting] .................... (3 of 4) Processing gauss, total=   0.0s
[Voting] ....................... (1 of 4) Processing rf, total=   0.8s
[Voting] ...................... (4 of 4) Processing SVM, total=  13.3s


In [255]:
temp = pd.DataFrame()

In [256]:
temp['id'] = kek[['record_id']]
temp['predict'] = pr

In [257]:
temp

id  predict
0       196        0
1      1196        0
2      2813        0
3      4385        0
4      6479        0
..      ...      ...
195  265577        0
196  266648        0
197  266849        0
198  267566        0
199  271584        0

[200 rows x 2 columns]

In [258]:
temp.to_csv('PD-submit.csv', index=False, sep=';')